enhancements/network/ovn-kubernetes-ipsec.md

```diff
@@ -0,0 +1,298 @@
+- All components used to implement this must be FIPs-compliant cryptographic
+components.
+- Provide an implementation that will be deployable on both Public and Private
+Cloud infrastructure. This will require that all inter-worker node traffic can
+be routed by the underlay network without reconfiguration of underlay firewall
+policies.
+- Reuse the same certificates, keys and CA used by the Cluster Management TLS
+implementation.
+
+### Non-Goals
+
+- Will not provide any Hardware Offload of the IPsec protocol. Although some
+checksums may be offloaded by standard NICs without any special enablement.
+- Will not provide any implementation for DPDK interfaces.
+- Will not provide an option to enable tunnels that have been configured with
+self-signed certificates or pre-shared keys.
+- Will not do any explicit performance optimization of the datapath. As the
+implementation of this enhancement will primarily require changes to the control
+and management path, it is expected that this will not alter the performance of
+the current OVS/Kernel IPsec implementation.
+- Will not provide any interface to allow a cluster operator to specify
+preferred encryption, integrity, authentication algorithms
+- Will not explicitly encrypt cluster management traffic. Cluster management
+traffic has already been encrypted using TLS.
+- Will not allow granular configuration of which nodes will have IPsec enabled
+(for example Namespace to Namespace encryption) or how IPsec is configured
+between specific nodes. This enhancement will enable IPsec between all worker
+nodes within the cluster indiscriminately.
+- Will not provide IPsec connectivity between any worker node that does not form
+part of the Kubernetes cluster. This enhancement is only for Node-to-Node
+(east-west) communication within the cluster.
+
+## Proposal
+
+The following changes are proposed to the [network.operator.openshift.io API](https://github.com/openshift/api/blob/master/operator/v1/types_network.go).
+
+```go
+// ovnKubernetesConfig contains the configuration parameters for networks
+// using the ovn-kubernetes network project
+type OVNKubernetesConfig struct {
+	// mtu is the MTU to use for the tunnel interface. This must be 100
+	// bytes smaller than the uplink mtu.
+	// Default is 1400
+	// +kubebuilder:validation:Minimum=0
+	// +optional
+	MTU *uint32 `json:"mtu,omitempty"`
+	// geneve port is the UDP port to be used by geneve encapulation.
+	// Default is 6081
+	// +kubebuilder:validation:Minimum=1
+	// +optional
+	GenevePort *uint32 `json:"genevePort,omitempty"`
+	// HybridOverlayConfig configures an additional overlay network for peers that are
+	// not using OVN.
+	// +optional
+  HybridOverlayConfig *HybridOverlayConfig `json:"hybridOverlayConfig,omitempty"`
+  // EnableIPsec enables IPsec encryption on all pod-pod traffic within the cluser. If unset, the plugin will default
+  // EnableIPsec to false.
+  // +optional
+  EnableIPsec *bool `json:"enableIPsec,omitempty"`
```

Comment:
ipsec has a bunch of configuration and operations model:
- agressive, normal and quick negotiation
- ikev2 vs ikev1
- keepalives
- hash and encryption algorithm
- key lifetime
...

if we don' t want to expose them at least we should define clearly what is and what is not supported

Reply:
@aojea Through documentation?

Reply:
yeah,  my experience with ipsec is that users and customers always ask these questions, also this information usually is required when companies are evaluation products.
From my perspective, since this is an OVN only feature is completely fine to say, ovn-kubernetes supports ipsec between node using ikev2, with crypto algorithms X, Y and Z , ....  just it should be documented

Reply:
Ok, I will see what I can do. It will realy be dependent on the libreswan/kernel version but I'll see if there is some way for me to summarize this.

Reply:
sorry, I thought you were going to hardcode some configurations parameters, I think that for some of them (the high level ones) you should,  depending on defaults can lead to problems on upgrade or nodes with skew versions if libreswan does not maintain compatibilities, 

i.e from https://libreswan.org/man/ipsec.conf.5.html
> ikev2

> Whether to use IKEv1 (RFC 4301) or IKEv2 (RFC 7296) settings to be used. Currently the accepted values are no(the default), signifying only IKEv1 is accepted, or yes, signifying only IKEv2 is accepted. Previous versions allowed the keywords propose or permit that would allow either IKEv1 or IKEv2, but this is no longer supported. The permit option is interpreted as no and the propose option is interpreted as yes. Older versions also supported keyword insist which is now interpreted as yes.

for the rest just mention something along those lines `will be dependant on the libreswan/kernel version ... `

Reply:
@aojea Ok, that makes sense. There is a similar comment in another place about "using sensible defaults".

Reply:
This will be provided as part of the documentation,